src/Instruction.cpp

```diff
@@ -38,34 +42,47 @@ uint8_t Instruction::getZERO(uint32_t inst)
     return (inst) ? 0x0 : 0x1;
 }
 
-uint8_t Instruction::getCarry(uint64_t result)
+uint32_t Instruction::addWithCarry(uint32_t op1, uint32_t op2, uint32_t carryInAsBit)
 {
-    uint16_t highBit = (result >> 32) & 0xffffffff;
+    int carryIn = carryInAsBit ? 1 : 0;
+    uint64_t unsignedSum = (uint64_t)op1 + (uint64_t)op2 + (uint64_t)carryIn;
+    int64_t signedSum = (int64_t)(int32_t)op1 + (int64_t)(int32_t)op2 + (int64_t)carryIn;
+    uint32_t result = (uint32_t)unsignedSum;
 
-    if(highBit != 0x00)
-        return 1;
-    else
-        return 0;
-}
-uint8_t Instruction::getBorrow(uint64_t result)
-{
-    if(((result >> 32) & 0x01) == 0x00)
-        return 1;
-    else
-        return 0;
+    reg->PSR[psrC] = ((uint64_t)result == unsignedSum) ? 0 : 1;
+    reg->PSR[psrV] = ((int64_t)(int32_t)result == signedSum) ? 0 : 1;
+
+    return result;
 }
-uint8_t Instruction::getOverflow(uint32_t op1, uint32_t op2, uint32_t result)
+
+uint32_t Instruction::shiftC(uint32_t value, uint8_t type, uint32_t shift, uint32_t carryIn)
 {
-    if(getMSB(op1) == getMSB(op2))
+    uint32_t result = value;
+    reg->PSR[psrC] = carryIn;
+
+    if (shift != 0)
     {
-        if(getMSB(op1) != getMSB(result))
-            return 1;
-        else
-            return 0;
+        switch (type)
+        {
+        case LSL:
```

Comment:
```cpp
value & (1 << (32 - shift));
(value & (1 << (shift - 1)));
reg->PSR[psrC] = result & (1 << 31);
```

I think, these might be exceed size of PSR[psrC], decleared with uint8_t .

```cpp
    uint32_t R[REG_SIZE];
    uint8_t PSR[PSR_SIZE];
```

The PSR should be either 0 or 1 (to represent false or true). 
in LSL, if value is 

1110_0110_0011_0000_1111_1001_0110_0011

and shift is 3, than result of "value & (1 << (32 - shift))" is 

0010_0000_0000_0000_0000_0000_0000_0000

I think, PSR[psrC] should store '1', not 0x20000000.

Was this result intentional? please check it before I merge this 
if not intentional, than I will fix and merge it